[Core] Do not report stats when worker is already dead.

rkooo567 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ffbin commented on src/ray/raylet/node_manager.cc:
```diff
@@ -3306,6 +3306,9 @@ void NodeManager::HandleGetNodeStats(const rpc::GetNodeStatsRequest &node_stats_
     return;
   }
   for (const auto &worker : all_workers) {
+    if (worker->IsDead()) {
```
Does it cause line3332(send_reply_callback(Status::OK(), nullptr, nullptr);) not to be called?

ffbin replied:
I prefer to add a default parameter to `GetAllRegisteredWorkers` and `GetAllRegisteredDrivers` to filter the worker and driver of alive (not filtered by default).

rkooo567 replied:
Sounds good. Let me fix this.

edoakes replied:
Shouldn't dead workers be filtered from those calls anyways? When do they actually get removed?

rkooo567 replied:
It doesn't look like it. It only returns "Registered" workers, and there seem to be cases where there are dead registered workers. Maybe when the actor is marked as dead and deregistered are different. After I introduced this line, I saw some core worker stats failures logs from raylet have disappeared.

rkooo567 commented:
Addressed comments @edoakes @ffbin 

For other functions, I feel like they could also set `filter_dead-*` to be true, but I just preserved the previous behavior just in case.

ffbin commented on src/ray/raylet/node_manager.cc:
```diff
@@ -496,8 +496,10 @@ void NodeManager::Heartbeat() {
-  auto all_workers = worker_pool_.GetAllRegisteredWorkers();
-  for (const auto &driver : worker_pool_.GetAllRegisteredDrivers()) {
+  auto all_workers =
+      worker_pool_.GetAllRegisteredWorkers(false /* filter_dead_workers */);
```
How about set a default value of filter_dead_workers? The previous implementation default is false.

rkooo567 replied:
Ah yeah. I tried, and it seemed to have a compiler error. But I think I just did it wrong. Is this the right way to do that? 

https://www.programiz.com/cpp-programming/default-argument

ffbin replied:
Yeah, this is the right way.

edoakes commented on src/ray/raylet/node_manager.cc:
```diff
@@ -496,8 +496,10 @@ void NodeManager::Heartbeat() {
-  auto all_workers = worker_pool_.GetAllRegisteredWorkers();
-  for (const auto &driver : worker_pool_.GetAllRegisteredDrivers()) {
+  auto all_workers =
+      worker_pool_.GetAllRegisteredWorkers(false /* filter_dead_workers */);
```
```suggestion
      worker_pool_.GetAllRegisteredWorkers(/* filter_dead_workers=*/false);
```
to stay consistent with other places in the code. same with below
